# Importing Data

## Loading and Displaying DataFrame



Importing the necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display, HTML
import warnings
warnings.filterwarnings("ignore")

Reading the CSV file into a dataFrame and displaying the first few rows of the DataFrame to examine its structure and contents

In [2]:
df = pd.read_csv('imputed_dataset.csv')
pd.set_option('display.max_columns', None)
df.head()

hospital_id   age        bmi  elective_surgery  ethnicity gender  height  \
0        118.0  68.0  22.732803               0.0  Caucasian      M   180.3   
1         81.0  77.0  27.421875               0.0  Caucasian      F   160.0   
2        118.0  25.0  31.952749               0.0  Caucasian      F   172.7   
3        118.0  81.0  22.635548               1.0  Caucasian      F   165.1   
4         33.0  19.0  29.110180               0.0  Caucasian      M   188.0   

  hospital_admit_source           icu_admit_source  icu_id icu_stay_type  \
0                 Floor                      Floor    92.0         admit   
1                 Floor                      Floor    90.0         admit   
2  Emergency Department       Accident & Emergency    93.0         admit   
3        Operating Room  Operating Room / Recovery    92.0         admit   
4  Emergency Department       Accident & Emergency    91.0         admit   

       icu_type  pre_icu_los_days     weight  albumin_apache  \
0         CTICU          0.541667  73.900000        2.300000   
1  Med-Surg ICU          0.927778  70.200000        2.886148   
2  Med-Surg ICU          0.000694  95.300000        2.886148   
3         CTICU          0.000694  61.700000        2.886148   
4  Med-Surg ICU          0.073611  83.792228        2.886148   

   apache_2_diagnosis  apache_3j_diagnosis  apache_post_operative  arf_apache  \
0               113.0               502.01                    0.0         0.0   
1               108.0               203.01                    0.0         0.0   
2               122.0               703.03                    0.0         0.0   
3               203.0              1206.03                    1.0         0.0   
4               119.0               601.01                    0.0         0.0   

   bilirubin_apache  bun_apache  creatinine_apache  gcs_eyes_apache  \
0          0.400000   31.000000           2.510000         3.000000   
1          1.201226    9.000000           0.560000         1.000000   
2          1.201226   25.717981           1.481635         3.000000   
3          1.201226   25.717981           1.481635         4.000000   
4          1.201226   25.717981           1.481635         3.488309   

   gcs_motor_apache  gcs_unable_apache  gcs_verbal_apache  glucose_apache  \
0          6.000000           0.000000           4.000000      168.000000   
1          3.000000           0.000000           1.000000      145.000000   
2          6.000000           0.000000           5.000000      160.136997   
3          6.000000           0.000000           5.000000      185.000000   
4          5.484908           0.011445           4.030254      160.136997   

   heart_rate_apache  hematocrit_apache  intubated_apache  map_apache  \
0              118.0          27.400000               0.0        40.0   
1              120.0          36.900000               0.0        46.0   
2              102.0          32.975752               0.0        68.0   
3              114.0          25.900000               1.0        60.0   
4               60.0          32.975752               0.0       103.0   

   resprate_apache  sodium_apache  temp_apache  urineoutput_apache  \
0             36.0     134.000000         39.3         1801.043127   
1             33.0     145.000000         35.1         1801.043127   
2             37.0     137.945522         36.7         1801.043127   
3              4.0     137.945522         34.8         1801.043127   
4             16.0     137.945522         36.7         1801.043127   

   ventilated_apache  wbc_apache  d1_diasbp_invasive_max  \
0                0.0   14.100000               46.000000   
1                1.0   12.700000               79.262762   
2                0.0   12.187859               79.262762   
3                1.0    8.000000               62.000000   
4                0.0   12.187859               79.262762   

   d1_diasbp_invasive_min  d1_diasbp_max  d1_diasbp_min  \
0               32.000000    

Checking the dimensions of the dataframe

In [3]:
df.shape

(130081, 131)

Re-evaluating for missing values

In [4]:
column_missing_info = []
for col in df.columns:
  if(df[col].isnull().any()):
    column_missing_info.append({
        'Column': col,
        'Data Type': df[col].dtype,
        'Null values count': df[col].isnull().sum(),
        '% of missing data': ((df[col].isnull().sum())/len(df))*100
    })

if column_missing_info:
  column_info_df = pd.DataFrame(column_missing_info)
  display(HTML(column_info_df.sort_values(by=["Data Type", "Null values count"]).to_html()))
else:
  print('No missing values!')

No missing values!


# Data Transformation

Dropping 'hospital_id', 'icu_id' columns as they didn't have any considerable correlation with the target variable

In [5]:
df = df.drop(columns=['hospital_id', 'icu_id'])
df.head()

age        bmi  elective_surgery  ethnicity gender  height  \
0  68.0  22.732803               0.0  Caucasian      M   180.3   
1  77.0  27.421875               0.0  Caucasian      F   160.0   
2  25.0  31.952749               0.0  Caucasian      F   172.7   
3  81.0  22.635548               1.0  Caucasian      F   165.1   
4  19.0  29.110180               0.0  Caucasian      M   188.0   

  hospital_admit_source           icu_admit_source icu_stay_type  \
0                 Floor                      Floor         admit   
1                 Floor                      Floor         admit   
2  Emergency Department       Accident & Emergency         admit   
3        Operating Room  Operating Room / Recovery         admit   
4  Emergency Department       Accident & Emergency         admit   

       icu_type  pre_icu_los_days     weight  albumin_apache  \
0         CTICU          0.541667  73.900000        2.300000   
1  Med-Surg ICU          0.927778  70.200000        2.886148   
2  Med-Surg ICU          0.000694  95.300000        2.886148   
3         CTICU          0.000694  61.700000        2.886148   
4  Med-Surg ICU          0.073611  83.792228        2.886148   

   apache_2_diagnosis  apache_3j_diagnosis  apache_post_operative  arf_apache  \
0               113.0               502.01                    0.0         0.0   
1               108.0               203.01                    0.0         0.0   
2               122.0               703.03                    0.0         0.0   
3               203.0              1206.03                    1.0         0.0   
4               119.0               601.01                    0.0         0.0   

   bilirubin_apache  bun_apache  creatinine_apache  gcs_eyes_apache  \
0          0.400000   31.000000           2.510000         3.000000   
1          1.201226    9.000000           0.560000         1.000000   
2          1.201226   25.717981           1.481635         3.000000   
3          1.201226   25.717981           1.481635         4.000000   
4          1.201226   25.717981           1.481635         3.488309   

   gcs_motor_apache  gcs_unable_apache  gcs_verbal_apache  glucose_apache  \
0          6.000000           0.000000           4.000000      168.000000   
1          3.000000           0.000000           1.000000      145.000000   
2          6.000000           0.000000           5.000000      160.136997   
3          6.000000           0.000000           5.000000      185.000000   
4          5.484908           0.011445           4.030254      160.136997   

   heart_rate_apache  hematocrit_apache  intubated_apache  map_apache  \
0              118.0          27.400000               0.0        40.0   
1              120.0          36.900000               0.0        46.0   
2              102.0          32.975752               0.0        68.0   
3              114.0          25.900000               1.0        60.0   
4               60.0          32.975752               0.0       103.0   

   resprate_apache  sodium_apache  temp_apache  urineoutput_apache  \
0             36.0     134.000000         39.3         1801.043127   
1             33.0     145.000000         35.1         1801.043127   
2             37.0     137.945522         36.7         1801.043127   
3              4.0     137.945522         34.8         1801.043127   
4             16.0     137.945522         36.7         1801.043127   

   ventilated_apache  wbc_apache  d1_diasbp_invasive_max  \
0                0.0   14.100000               46.000000   
1                1.0   12.700000               79.262762   
2                0.0   12.187859               79.262762   
3                1.0    8.000000               62.000000   
4                0.0   12.187859               79.262762   

   d1_diasbp_invasive_min  d1_diasbp_max  d1_diasbp_min  \
0               32.000000           68.0           37.0   
1               46.970187           95.0           31.0   
2               46.970187           88

Separating and Storing numerical and categorical columns

In [6]:
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = df.select_dtypes(include=['object']).columns

Taking copy of orginal dataframe before making any transformations

In [7]:
df_original = df.copy(deep=True)

## Categorical encoding

Performing one-hot encoding on categorical variables and assigning the new encoded columns to the dataframe

In [8]:
discret_df = pd.get_dummies(df[categorical_cols], drop_first = True)
df = pd.concat([df, discret_df], axis=1)
df.drop(categorical_cols, axis=1, inplace=True)
df.replace({True: 1, False: 0}, inplace=True)

display(HTML(df.head().to_html()))

Separating numeric columns from binary columns, so that standardization can be applied to numeric columns only

In [9]:
binary_columns = [x for x in df.columns if (df[x].nunique() == 2 and set(df[x].unique()) == {0, 1})]
numeric_cols = [x for x in df.columns if x not in binary_columns]

## Performing Data Normalization

Performing Data Normalization to ensure that all numerical features in the dataset have a similar scale

In [10]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df[numeric_cols] = pd.DataFrame(scaler.fit_transform(df[numeric_cols]), columns=numeric_cols)
display(HTML(df.head().to_html()))

# Model Training and Evaluation

Separating the target feature from the independent features for model training

In [11]:
y_feature = df['diabetes_mellitus']
x_features = df.drop('diabetes_mellitus',axis=1)

## Model - PCA

In [12]:
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from imblearn.over_sampling import SMOTE
from sklearn.decomposition import PCA

# Iterating over PCA values with a gap of 10
for n_components in range(7, 97, 10):
    print(f"Applying PCA with {n_components} components")
    pca = PCA(n_components=n_components)
    x_features_After_pca = pca.fit_transform(x_features)

    # Splitting the data into train and test sets
    x_train, x_test, y_train, y_test = train_test_split(x_features_After_pca, y_feature, test_size=0.2, random_state=42)

    # Balancing dataset using SMOTE to resolve imbalance Issue
    sm = SMOTE(random_state=42)
    x_resampled, y_resampled = sm.fit_resample(x_train, y_train)

    # Training the XGBoost Classifier
    model = XGBClassifier(
        n_estimators=1500,
        max_depth=3,
        learning_rate=0.1,
        gamma=0,
        min_child_weight=1,
        subsample=0.8,
        colsample_bytree=0.8,
        reg_alpha=0,
        reg_lambda=1
    )
    model.fit(x_resampled, y_resampled)

    # Evaluating the model on the test set
    y_pred = model.predict(x_test)
    model_accuracy = accuracy_score(y_test, y_pred)
    model_precision = precision_score(y_test, y_pred, average='macro')
    model_recall = recall_score(y_test, y_pred, average='macro')
    model_f1 = f1_score(y_test, y_pred, average='macro')
    model_auc = roc_auc_score(y_test, y_pred)

    print('Model performance (Hold-out Validation)')
    print("- Accuracy: {:.4f}".format(model_accuracy))
    print("- Precision: {:.4f}".format(model_precision))
    print("- Recall: {:.4f}".format(model_recall))
    print("- F1-Score: {:.4f}".format(model_f1))
    print("- AUC: {:.4f}".format(model_auc))
    print('='*35)
    print('\n')

Applying PCA with 7 components
Model performance (Hold-out Validation)
- Accuracy: 0.5635
- Precision: 0.5441
- Recall: 0.5644
- F1-Score: 0.5146
- AUC: 0.5644


Applying PCA with 17 components
Model performance (Hold-out Validation)
- Accuracy: 0.6571
- Precision: 0.5999
- Recall: 0.6398
- F1-Score: 0.5947
- AUC: 0.6398


Applying PCA with 27 components
Model performance (Hold-out Validation)
- Accuracy: 0.7408
- Precision: 0.6602
- Recall: 0.7053
- F1-Score: 0.6707
- AUC: 0.7053


Applying PCA with 37 components
Model performance (Hold-out Validation)
- Accuracy: 0.7468
- Precision: 0.6634
- Recall: 0.7058
- F1-Score: 0.6745
- AUC: 0.7058


Applying PCA with 47 components
Model performance (Hold-out Validation)
- Accuracy: 0.7537
- Precision: 0.6680
- Recall: 0.7079
- F1-Score: 0.6795
- AUC: 0.7079


Applying PCA with 57 components
Model performance (Hold-out Validation)
- Accuracy: 0.7610
- Precision: 0.6725
- Recall: 0.7084
- F1-Score: 0.6841
- AUC: 0.7084


Applying PCA with 67 co

Comparatively, PCA with 47 components is the best PCA model.
It is noticed that despite increasing the PCA components beyond 47, the increase in model performance is negligible as we are using more PCA's for little gain.

## XGBoost with PCA - Hold Out Evaluation

In [13]:
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from imblearn.over_sampling import SMOTE
from sklearn.decomposition import PCA


print(f"Applying PCA with 47 components")
pca = PCA(n_components=47)
x_features_After_pca = pca.fit_transform(x_features)

# Splitting the data into train and test sets
x_train, x_test, y_train, y_test = train_test_split(x_features_After_pca, y_feature, test_size=0.2, random_state=42)

# Balancing dataset using SMOTE to resolve imbalance Issue
sm = SMOTE(random_state=42)
x_resampled, y_resampled = sm.fit_resample(x_train, y_train)

# Training the XGBoost Classifier
model = XGBClassifier(
    n_estimators=2500,
    max_depth=3,
    learning_rate=0.1,
    gamma=0.75,
    min_child_weight=1,
    subsample=0.8,
    colsample_bytree=0.9,
    reg_alpha=0,
    reg_lambda=1
)
model.fit(x_resampled, y_resampled)

# Evaluating the model on the test set
y_pred = model.predict(x_test)
model_accuracy = accuracy_score(y_test, y_pred)
model_precision = precision_score(y_test, y_pred, average='macro')
model_recall = recall_score(y_test, y_pred, average='macro')
model_f1 = f1_score(y_test, y_pred, average='macro')
model_auc = roc_auc_score(y_test, y_pred)

print("XGBoost Classifier by using best PCA components")
print('Model performance (Hold-out Validation)')
print("- Accuracy: {:.4f}".format(model_accuracy))
print("- Precision: {:.4f}".format(model_precision))
print("- Recall: {:.4f}".format(model_recall))
print("- F1-Score: {:.4f}".format(model_f1))
print("- AUC: {:.4f}".format(model_auc))
print('='*35)
print('\n')

Applying PCA with 47 components
XGBoost Classifier by using best PCA components
Model performance (Hold-out Validation)
- Accuracy: 0.7563
- Precision: 0.6679
- Recall: 0.7041
- F1-Score: 0.6792
- AUC: 0.7041




The above XGBoost (Hold-out) model has an accuracy of 75.63%
The F1 score which is a function of precision and recall metrics, has a value of 67.92%
The AUC value which determines the ability of the model to distingusish between positive and negative labels is 70.41%
Hence, the XGBoost (Hold-out) model has reasonably good prediction accuracy.

## XGBoost with PCA - N fold Cross validation

Perform N fold Cross validation on 47 PCA components obtained

In [16]:
from sklearn.model_selection import cross_val_predict, KFold
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from imblearn.over_sampling import SMOTE

# Create a 5-fold cross-validation object
kfold = KFold(n_splits=5, shuffle=True, random_state=42)

data_5folds = []

accuracy_list = []
precision_list = []
recall_list = []
f1_list = []
auc_list = []

pca = PCA(n_components=47)
x_features_After_pca = pca.fit_transform(x_features)

for train_index, test_index in kfold.split(x_features_After_pca,y_feature):
    # get actual data by index
    x_train, x_test = x_features.iloc[train_index], x_features.iloc[test_index]
    y_train, y_test = y_feature[train_index], y_feature[test_index]
    # save data into fold
    fold = [x_train, x_test, y_train, y_test]
    # add each fold data into 5folds
    data_5folds.append(fold)

for x_train, x_test, y_train, y_test in data_5folds:
    sm = SMOTE(random_state=42)
    sm.fit(x_train, y_train)
    x_resampled, y_resampled = sm.fit_resample(x_train, y_train)
    model = XGBClassifier(
        n_estimators=1500,
        max_depth=3,
        learning_rate=0.1,
        gamma=0,
        min_child_weight=1,
        subsample=0.8,
        colsample_bytree=0.8,
        reg_alpha=0,
        reg_lambda=1
        )
    model.fit(x_resampled, y_resampled)
    y_pred = model.predict(x_test)
    # Calculate the performance metrics
    model_accuracy = accuracy_score(y_test, y_pred)
    model_precision = precision_score(y_test, y_pred, average='macro')
    model_recall = recall_score(y_test, y_pred, average='macro')
    model_f1 = f1_score(y_test, y_pred, average='macro')
    model_auc = roc_auc_score(y_test, y_pred)
    
    accuracy_list.append(model_accuracy)
    precision_list.append(model_precision)
    recall_list.append(model_recall)
    f1_list.append(model_f1)
    auc_list.append(model_auc)

print('XGBoost Classifier Model performance(5-Fold CV)')
print("- Accuracy: {:.4f}".format(np.mean(accuracy_list)))
print("- Precision: {:.4f}".format(np.mean(precision_list)))
print("- Recall: {:.4f}".format(np.mean(recall_list)))
print("- F1-Score: {:.4f}".format(np.mean(f1_list)))
print("- AUC: {:.4f}".format(np.mean(auc_list)))
print('='*35)
print('\n')

XGBoost Classifier Model performance(5-Fold CV)
- Accuracy: 0.8310
- Precision: 0.7564
- Recall: 0.7056
- F1-Score: 0.7247
- AUC: 0.7056




The accuracy (83.10%) of the above XGBoost Classifier (5-Fold CV) is the best among all other models. Particularily, the ability of the model to identify the diabetes_mellitus positive labels is high.

However, the lack of model performance with regards to the AUC value (70.56%) is evident. It indicates the shortcomings of the model to distinguish between positive and negative labels.

Overall, The above model is amongst our best performing models